New usdtree command line tool

aloysbaillet opened the pull request:
### Description of Change(s)
Added a new `usdtree` command line tool that is very similar to Alembic's abctree.

```bash
usage: usdtree [-h] [--unloaded] [--attributes] [--metadata] [--simple]
               [--flatten] [--flattenLayerStack]
               [--mask PRIMPATH[,PRIMPATH...]]
               inputPath

Writes the tree structure of a USD file. The default is to inspect a single
USD file. Use the --flatten argument to see the flattened (or composed) Stage
tree. Special metadata "kind" and "active" are always shown if authored unless
--simple is provided.

positional arguments:
  inputPath

optional arguments:
  -h, --help            show this help message and exit
  --unloaded            Do not load payloads
  --attributes, -a      Display authored attributes
  --metadata, -m        Display authored metadata (active and kind are part of
                        the label and not shown as individual items)
  --simple, -s          Only display prim names: no specifier, kind or active
                        state.
  --flatten, -f         Compose the stage with the input file as root layer
                        and write the flattened content.
  --flattenLayerStack   Flatten the layer stack with the given root layer.
                        Unlike --flatten, this does not flatten composition
                        arcs (such as references).
  --mask PRIMPATH[,PRIMPATH...]
                        Limit stage population to these prims, their
                        descendants and ancestors. To specify multiple paths,
                        either use commas with no spaces or quote the argument
                        and separate paths by commas and/or spaces. Requires
                        --flatten.
```

Example output:
```
usdtree testUsdTreePayload/input.usda --flatten
/
 `--root [def Scope] (kind = assembly)
     |--child1 [def Xform] (kind = group)
     |   |--subchild1a [def Cube]
     |   `--subchild1b [def Cube]
     `--child2 [def Xform] (kind = group)
         |--subchild2a [def Xform] (kind = component)
         |   `--subgrandchild2a1 [def Cube] (kind = subcomponent)
         `--subchild2b [def Xform] (active = false, kind = component)
```

jtran56 commented:
Filed as internal issue #USD-4740.

spiffmon commented on pxr/usd/bin/usdtree/usdtree.py:
```diff
@@ -0,0 +1,228 @@
+                stage = Usd.Stage.Open(path, Usd.Stage.LoadNone)
+            else:
+                stage = Usd.Stage.Open(path)
+        if args.flattenLayerStack:
```
A neat idea to try to combine population masking with layerStack flattening! Unfortunately, our doc is a bit misleading for UsdUtilsFlattenLayerStack: it doesn't use the input stage in any meaningful way - it only grabs the root layerStack out of it, which is unaffected by population masking.
Further, in other tools, --flatten and --flattenLayerStack are independent command-line args, so there's no reason to make them dependent here.

And when you open a Stage in the flattenLayerStack case, you can *always* open it unloaded, because load-state also does not affect the root layerStack.

spiffmon commented on pxr/usd/bin/usdtree/usdtree.py:
```diff
@@ -0,0 +1,228 @@
+                stage = Usd.Stage.Open(path)
+        if args.flattenLayerStack:
+            from pxr import UsdUtils
+            stage = UsdUtils.FlattenLayerStack(stage)
```
UsdUtils.FlattenLayerStack returns a Layer, not a Stage.

spiffmon commented on pxr/usd/bin/usdtree/usdtree.py:
```diff
@@ -0,0 +1,228 @@
+        if args.flattenLayerStack:
+            from pxr import UsdUtils
+            stage = UsdUtils.FlattenLayerStack(stage)
+        PrintStage(args, stage)
```
You _could_ do away with PrintStage altogether by using the stage.Flatten() method to get yourself a Layer, which would simplify the code a bunch, and ensure more consistent behavior (like the note about GetChildren() above, and also ensure that all "masters" get presented in some form when there's instancing).

It will be a little slower, but shouldn't be too bad if all the inputs are USD files, because we won't pull in any of the expensive attribute data, but might be quite a bit slower for alembic files.  I could see going either way - just wanted to present the options.

spiffmon commented on pxr/usd/bin/usdtree/usdtree.py:
```diff
@@ -0,0 +1,228 @@
+
+
+if __name__ == "__main__":
+    # Restore signal handling defaults to allow output redirection and the like.
```
Nice!

spiffmon commented on pxr/usd/bin/usdtree/usdtree.py:
```diff
@@ -0,0 +1,228 @@
+        help='Display prim types')
+    parser.add_argument(
+        '-f', '--flatten', action='store_true', help='Compose stages with the '
+        'input files as root layers and write their flattened content')
```
inputFile is a singular input, so we should talk about 'stage' and 'layer' above, not 'stages' and 'layers'

spiffmon commented:
One more thing, @aloysbaillet - it might be useful to give `kind` metadata special treatment, also, actually printing its value, as it'll always be a short string, and as per a recent usd-interest thread, may aid in debugging model-hierarchy

aloysbaillet commented:
Thanks a lot @spiffmon ! I'll try and have a look at this next week. I'm now quite tempted to use the Flatten method as I didn't realise it would simplify the code so much...

aloysbaillet commented:
So I've cleaned the logic to match `usdcat` better, and replaced `--types` by `--definitions`, and added `--kinds`, and also added ` (inactive)` if a prim is inactive...
New output looks like:
```
usdtree testUsdTreePayload/input.usda --definitions --kinds
USD
 `--def root [kind=assembly]
     |--def child1 [kind=group]
     `--def child2 [kind=group]
         |--def subchild2a [kind=component]
         |   `--def subgrandchild2a1 [kind=subcomponent]
         `--def subchild2b [kind=component] (inactive)
```

Now I'm kinda struggling to choose the best default values so the tool is as useful out of the box. I'm tempted to make `--definitions` and `--kinds` the default.

Any opinion @spiffmon ?

Also, I'm quite tempted to add metadata and attribute values... but that might be a bit of a rabbit hole...

spiffmon commented:
I'd agree let's maybe keep this simple and easy for now.  One of the things I find principally useful about a usdtree tool is how easy it makes reading the namespace hierarchy.  So, although output like 
`--- def Xform child1`
looks more like usda, it makes it harder to parse the hierarchy, because your eye needs to hop over a variable amount of words and chars on each line to find the parent.  So I'd argue to put *all* extra info _after_ the prim name, inside brackets or parentheses.  Maybe
`--- child1 [def Xform] (active = true, kind = component)`

Note also what I did with _active_ there: because this is primarily a *layer* inspection tool, it is useful information to know if _active_ is authored (and to what value) rather than just telling whether it is inactive.  You can use UsdPrim.HasAuthoredActive() to tell if it's been authored at all, for that API.

In terms of options, maybe merge types, definitions, kinds into a single "primData" option?  Default would still just be simply the namespace.

Codewise, don't you want to include _active_ and _kind_ in the METADATA_TO_SKIP list?

aloysbaillet commented:
Thanks @spiffmon !
I'm still torn about the default, I went the other way for now:
```
usdtree testUsdTreePayload/input.usda --flatten
USD
 `--root [def Scope] (kind = assembly)
     |--child1 [def Xform] (kind = group)
     |   |--subchild1a [def Cube]
     |   `--subchild1b [def Cube]
     `--child2 [def Xform] (kind = group)
         |--subchild2a [def Xform] (kind = component)
         |   `--subgrandchild2a1 [def Cube] (kind = subcomponent)
         `--subchild2b [def Xform] (active = false, kind = component)
```
and
```
usdtree testUsdTreePayload/input.usda --flatten --simple
USD
 `--root
     |--child1
     |   |--subchild1a
     |   `--subchild1b
     `--child2
         |--subchild2a
         |   `--subgrandchild2a1
         `--subchild2b
```
I feel having the specifier, type, kind and active=false seems quite useful by default... But I'm quite happy to revert and make simple the default and add `primData` as an option instead!

spiffmon commented:
I think your default makes sense, Aloys - let’s see if anyone else lurking
here wants to register an opinion.

On Tue, Feb 12, 2019 at 7:28 PM Aloys Baillet <notifications@github.com>
wrote:

> Thanks @spiffmon <https://github.com/spiffmon> !
> I'm still torn about the default, I went the other way for now:
>
> usdtree testUsdTreePayload/input.usda --flatten
> USD
>  `--root [def Scope] (kind = assembly)
>      |--child1 [def Xform] (kind = group)
>      |   |--subchild1a [def Cube]
>      |   `--subchild1b [def Cube]
>      `--child2 [def Xform] (kind = group)
>          |--subchild2a [def Xform] (kind = component)
>          |   `--subgrandchild2a1 [def Cube] (kind = subcomponent)
>          `--subchild2b [def Xform] (active = false, kind = component)
>
> and
>
> usdtree testUsdTreePayload/input.usda --flatten --simple
> USD
>  `--root
>      |--child1
>      |   |--subchild1a
>      |   `--subchild1b
>      `--child2
>          |--subchild2a
>          |   `--subgrandchild2a1
>          `--subchild2b
>
> I feel having the specifier, type, kind and active=false seems quite
> useful by default... But I'm quite happy to revert and make simple the
> default and add primData as an option instead!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463042628>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7qaIzPzA7Zfw9BuPDRihhYQ55OpZAUks5vM4ZegaJpZM4WouWF>
> .
>
-- 
--spiffiPhone

spitzak commented:
I think it could be shortened to remove repetitive elements. Instead of
"[def Scope]" just have "Scope" (use "over" and "class" for non-def, the
assumption is that there is no class actually called "over" or "class").
Also instead of "(kind = component)" just have "component". I would also
remove the top-level "USD".


On Tue, Feb 12, 2019 at 9:12 PM F. Sebastian (spiff) Grassia <
notifications@github.com> wrote:

> I think your default makes sense, Aloys - let’s see if anyone else lurking
> here wants to register an opinion.
>
> On Tue, Feb 12, 2019 at 7:28 PM Aloys Baillet <notifications@github.com>
> wrote:
>
> > Thanks @spiffmon <https://github.com/spiffmon> !
> > I'm still torn about the default, I went the other way for now:
> >
> > usdtree testUsdTreePayload/input.usda --flatten
> > USD
> > `--root [def Scope] (kind = assembly)
> > |--child1 [def Xform] (kind = group)
> > | |--subchild1a [def Cube]
> > | `--subchild1b [def Cube]
> > `--child2 [def Xform] (kind = group)
> > |--subchild2a [def Xform] (kind = component)
> > | `--subgrandchild2a1 [def Cube] (kind = subcomponent)
> > `--subchild2b [def Xform] (active = false, kind = component)
> >
> > and
> >
> > usdtree testUsdTreePayload/input.usda --flatten --simple
> > USD
> > `--root
> > |--child1
> > | |--subchild1a
> > | `--subchild1b
> > `--child2
> > |--subchild2a
> > | `--subgrandchild2a1
> > `--subchild2b
> >
> > I feel having the specifier, type, kind and active=false seems quite
> > useful by default... But I'm quite happy to revert and make simple the
> > default and add primData as an option instead!
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463042628
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AF7qaIzPzA7Zfw9BuPDRihhYQ55OpZAUks5vM4ZegaJpZM4WouWF
> >
> > .
> >
> --
> --spiffiPhone
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463061235>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABLBER8YMdjI54hOUmhFeOENhYklOFFxks5vM57bgaJpZM4WouWF>
> .
>

spiffmon commented:
I'm not sure about eliding "def"... is that actually the most common?  I'd
think "over" might be??

But I really like the explicitness of "(kind = component)" for two reasons:
1. It reinforces that *kind* is the metadatum that encodes the concept
2. I've been looking at this from a "pipe to grep" perspective, and you
might have prims named "component" or something like that, whereas grepping
for "kind = component" is guaranteed to find you just the component model
prims.

--spiff


On Wed, Feb 13, 2019 at 12:30 PM Bill Spitzak <notifications@github.com>
wrote:

> I think it could be shortened to remove repetitive elements. Instead of
> "[def Scope]" just have "Scope" (use "over" and "class" for non-def, the
> assumption is that there is no class actually called "over" or "class").
> Also instead of "(kind = component)" just have "component". I would also
> remove the top-level "USD".
>
>
> On Tue, Feb 12, 2019 at 9:12 PM F. Sebastian (spiff) Grassia <
> notifications@github.com> wrote:
>
> > I think your default makes sense, Aloys - let’s see if anyone else
> lurking
> > here wants to register an opinion.
> >
> > On Tue, Feb 12, 2019 at 7:28 PM Aloys Baillet <notifications@github.com>
> > wrote:
> >
> > > Thanks @spiffmon <https://github.com/spiffmon> !
> > > I'm still torn about the default, I went the other way for now:
> > >
> > > usdtree testUsdTreePayload/input.usda --flatten
> > > USD
> > > `--root [def Scope] (kind = assembly)
> > > |--child1 [def Xform] (kind = group)
> > > | |--subchild1a [def Cube]
> > > | `--subchild1b [def Cube]
> > > `--child2 [def Xform] (kind = group)
> > > |--subchild2a [def Xform] (kind = component)
> > > | `--subgrandchild2a1 [def Cube] (kind = subcomponent)
> > > `--subchild2b [def Xform] (active = false, kind = component)
> > >
> > > and
> > >
> > > usdtree testUsdTreePayload/input.usda --flatten --simple
> > > USD
> > > `--root
> > > |--child1
> > > | |--subchild1a
> > > | `--subchild1b
> > > `--child2
> > > |--subchild2a
> > > | `--subgrandchild2a1
> > > `--subchild2b
> > >
> > > I feel having the specifier, type, kind and active=false seems quite
> > > useful by default... But I'm quite happy to revert and make simple the
> > > default and add primData as an option instead!
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463042628
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AF7qaIzPzA7Zfw9BuPDRihhYQ55OpZAUks5vM4ZegaJpZM4WouWF
> > >
> > > .
> > >
> > --
> > --spiffiPhone
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463061235
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABLBER8YMdjI54hOUmhFeOENhYklOFFxks5vM57bgaJpZM4WouWF
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463356457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7qaF8bLhdXQC7g3RpyaYc4_pmG5Uxwks5vNHX9gaJpZM4WouWF>
> .
>

aloysbaillet commented:
To remove the top-level USD I would have to replace it with a pseudo-root somehow... and because we actually use "root" as the root prim in many layers I'd rather not call it "root" like usdview does... but I don't mind if you prefer I could just remove it completely as well!

spiffmon commented:
I'll note two things, without expressing an opinion:
1. The pseudoroot does have a name, it's "/"
2. On linux, `tree / | head` produces:
```
/
|-- bin -> usr/bin
|-- boot
|   |-- config-3.10.0-862.3.2.el7.px.x86_64
|   |-- efi [error opening dir]
|   |-- elf-memtest86+-5.01
|   |-- grub2 [error opening dir]

```

aloysbaillet commented:
Thanks everyone!
I like your non-opinionated note Spiff, and things now look like this:
```
usdtree testUsdTreePayload/input.usda --flatten
/
 `--root [def Scope] (kind = assembly)
     |--child1 [def Xform] (kind = group)
     |   |--subchild1a [def Cube]
     |   `--subchild1b [def Cube]
     `--child2 [def Xform] (kind = group)
         |--subchild2a [def Xform] (kind = component)
         |   `--subgrandchild2a1 [def Cube] (kind = subcomponent)
         `--subchild2b [def Xform] (active = false, kind = component)
```

drwave commented:
This is really cool...

> On Feb 14, 2019, at 7:45 PM, Aloys Baillet <notifications@github.com> wrote:
> 
> Thanks everyone!
> I like your non-opinionated note Spiff, and things now look like this:
> 
> usdtree testUsdTreePayload/input.usda --flatten
> /
>  `--root [def Scope] (kind = assembly)
>      |--child1 [def Xform] (kind = group)
>      |   |--subchild1a [def Cube]
>      |   `--subchild1b [def Cube]
>      `--child2 [def Xform] (kind = group)
>          |--subchild2a [def Xform] (kind = component)
>          |   `--subgrandchild2a1 [def Cube] (kind = subcomponent)
>          `--subchild2b [def Xform] (active = false, kind = component)
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-463895524>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABmgOniesJCkSHEwIOhpRjl2O03eg7daks5vNi1sgaJpZM4WouWF>.
>

spiffmon commented:
Hi @aloysbaillet !
   Just had a chance to review in perforce (which is easier than on gitHub) and noted just two last, small things:
1. The change description does not reflect the final command-line args when it describes usage
2. In your handling of "active", you are still not checking whether "active" is authored.  I am only harping on this because in some scenario, knowing that a prim has authored "active = True" is going to be important to someone in some debugging session.  For UsdPrim it's HasAuthoredActive(), and for SdfPrimSpec it's HasActive()

Cheers,
--spiff

aloysbaillet commented:
Thanks for the reminder @spiffmon I just added the printing of "active = true" if it's authored.
I'm not checking if active is authored if it is false, but I'm assuming that it is not possible to have active=false if it is not authored?
I also updated the PR description to match the new output!

spiffmon commented:
Correct on active=false, Aloys, and thanks for turning around the changes
so fast!

On Tue, Feb 19, 2019 at 6:58 PM Aloys Baillet <notifications@github.com>
wrote:

> Thanks for the reminder @spiffmon <https://github.com/spiffmon> I just
> added the printing of "active = true" if it's authored.
> I'm not checking if active is authored if it is false, but I'm assuming
> that it is not possible to have active=false if it is not authored?
> I also updated the PR description to match the new output!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/USD/pull/619#issuecomment-465400982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7qaOSYG31UpVxWhDKumDRzUPP9UYc9ks5vPLnPgaJpZM4WouWF>
> .
>
-- 
--spiffiPhone
